patchy/api.py

```diff
@@ -148,25 +149,74 @@ def _get_source(func):
         return source
 
 
-def _set_source(func, new_source):
+def _class_name(func):
+    qualname = getattr(func, '__qualname__', None)
+    if qualname is not None:
+        split_name = qualname.split('.')
+        try:
+            class_name = split_name[-2]
+        except IndexError:
+            return None
+        else:
+            if class_name == '<locals>':
+                return None
+            return class_name
+    im_class = getattr(func, 'im_class', None)
+    if im_class is not None:
+        return im_class.__name__
+
+
+def _set_source(func, func_source):
     # Fetch the actual function we are changing
     real_func = _get_real_func(func)
-
     # Figure out any future headers that may be required
     feature_flags = real_func.__code__.co_flags & FEATURE_MASK
 
+    def _compile(code, flags=0):
+        return compile(
+            code,
+            '<patchy>',
+            'exec',
+            flags=feature_flags | flags,
+            dont_inherit=True,
+        )
+
+    def _parse(code):
+        return _compile(code, flags=ast.PyCF_ONLY_AST)
+
+    class_name = _class_name(func)
```

Comment:
there is so much going on here now, i think we need it splitting into more functions and some comments added